core/solver/idr.cpp

```diff
@@ -0,0 +1,261 @@
+
+    auto exec = this->get_executor();
+
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+    auto dense_b = as<Vector>(b);
+    auto dense_x = as<Vector>(x);
+
+    constexpr uint8 RelativeStoppingId{1};
+
+    const auto problem_size = system_matrix_->get_size()[0];
+    const auto nrhs = dense_b->get_size()[1];
+
+    auto residual = Vector::create_with_config_of(dense_b);
+    auto v = Vector::create_with_config_of(dense_b);
+    auto t = Vector::create_with_config_of(dense_b);
+    auto helper = Vector::create_with_config_of(dense_b);
+
+    auto m =
+        Vector::create(exec, gko::dim<2>{subspace_dim_, subspace_dim_ * nrhs});
+
+    auto g =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+    auto u =
+        Vector::create(exec, gko::dim<2>{problem_size, subspace_dim_ * nrhs});
+
+    auto f = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+    auto c = Vector::create(exec, gko::dim<2>{subspace_dim_, nrhs});
+
+    auto omega = Vector::create(exec, gko::dim<2>{1, nrhs});
+    auto residual_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto tht = Vector::create(exec, dim<2>{1, nrhs});
+    auto t_norm = NormVector::create(exec, dim<2>{1, nrhs});
+    auto alpha = Vector::create(exec, gko::dim<2>{1, nrhs});
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(exec, nrhs);
+
+    auto subspace_vectors_ =
+        Vector::create(exec, gko::dim<2>(subspace_dim_, problem_size));
+
+    // Initialization
+    // m = identity
+    exec->run(idr::make_initialize(m.get(), subspace_vectors_.get(),
+                                   deterministic_, &stop_status));
+
+    // omega = 1
+    exec->run(
+        idr::make_fill_array(omega->get_values(), nrhs, one<ValueType>()));
+
+    // residual = b - Ax
+    residual->copy_from(dense_b);
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(),
+                          residual.get());
+
+    // g = u = 0
+    exec->run(idr::make_fill_array(
+        g->get_values(), problem_size * g->get_stride(), zero<ValueType>()));
+    exec->run(idr::make_fill_array(
+        u->get_values(), problem_size * u->get_stride(), zero<ValueType>()));
+
+
+    auto stop_criterion = stop_criterion_factory_->generate(
+        system_matrix_, std::shared_ptr<const LinOp>(b, [](const LinOp *) {}),
+        x, residual.get());
+
+    int total_iter = -1;
+
+    while (true) {
+        ++total_iter;
+        this->template log<log::Logger::iteration_complete>(
+            this, total_iter, residual.get(), dense_x);
+
+        if (stop_criterion->update()
+                .num_iterations(total_iter)
+                .residual(residual.get())
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        subspace_vectors_->apply(residual.get(), f.get());
```

Comment:
Maybe I am wrong, but I thought subspace_vectors only be transposed not conjugated.

Reply:
True, but as they are randomly generated, I decided to just generate P^H and skip the conjugating.